### ISSUE TITLE
Enhance Grid-Diag to compute mutual information

### ISSUE THREAD

**Issue by JohnHalleyGotway** [dtcenter/MET#3171]
## Describe the Enhancement ##
This work is requested for the METplus Land Projects to enhance MET to compute mutual information (MI) between two variables, such as 2 m temperature and top layer soil moisture. MI is a straightforward matrix operation, and the https://github.com/madvn/infotheory repo has test cases and C++ code to compute it.

Specifically: https://github.com/madvn/infotheory/blob/master/infotheory/InfoTools.h line 385 starts the mutual information function
Test case for MI is in python using the C++ function: https://github.com/madvn/infotheory/blob/master/benchmarks/mi_baseline.py

Recommend enhancing Grid-Diag to compute MI. Grid-Diag writes a NetCDF output file containing 1D and 2D histograms of the binned data values. Still need to determine if/how the MI results should be written to the existing output file. Or should a new output type be added? Should MI only be computed by Grid-Diag or should it be added to Point-Stat, Grid-Stat, and/or Series-Analysis? 

### Time Estimate ###
About a week

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
7750034 UAlb NOAA METplus UFS

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Review default **alert** labels
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Milestone and Projects ###
- [x] Select **Milestone** as a **MET-X.Y.Z** version, **Consider for Next Release**, or **Backlog of Development Ideas**
- [x] For a **MET-X.Y.Z** version, select the **MET-X.Y Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- [x] dtcenter/METplus#3194 to update the METplus Grid-Diag wrapper.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y Development** project for development toward the next coordinated release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

**Comment by anewman89** [comment]
@JohnHalleyGotway We may need to have a discussion on how to output MI and where computations could reside.  Best case scenario, MI can be computed from point observations and compared to nearest model grid cell using a time series of data, and also on gridded obs and model data.  

However, MI is more of a boutique metric right now, and it may be appropriate to keep it in a contained tool. It also seems like MI computation in Point-Stat, Grid-Stat, and Series-Analysis would require substantial modifications to those tools?

**Comment by JohnHalleyGotway** [comment]
@JohnHalleyGotway and @anewman89 met on Aug 1 to discuss and note the following:
- Mutual information should be computed with the same inputs used to compute the 2D histograms.
- Consider adding a config option to control what output is computed, e.g. 1D hist, 2D hist, MI.
- @anewman89 will consider whether we should support collapsing data from multiple times into one output. e.g. `collapse_times = true/false` [false by default]. But details remain to be defined.
- Recommend using the synthetic data examples from [madvn/infotheory](https://github.com/madvn/infotheory) to validate the implementation in Grid-Diag.
- Other stats exist in the realm of information theory and could be provided in the output with MI, including entropy values of the single input variables. Also look for directional estimators in the output. Consider adding these to the output as well.

Recommend starting with computing entropy and MI.
Going forward, consider adding support for adding support for partial information for multiple variables, adding support for multiple input times, and explore [directional estimators](https://en.wikipedia.org/wiki/Directed_information) (not included in the infotheory package).

This could also be implemented in Stat-Analysis so it can be computed on MPR inputs, but this is somewhat tricky. For example, we'd use soil moisture and 2-m temperature MPR data. We'd need to homogenize the data to the common cases and then extract time series of forecast and observation values. Recommend splitting this into a separate issue. Also recommend supporting time lags when computing MI.

This could also be implemented in Series-Analysis on a sufficiently long time series of data and compute MI for each individual grid point to see how MI changes spatially.

**Comment by JohnHalleyGotway** [comment]
Reassigning this issue from the METplus-6.2 project board to METplus-7.0 since we have the flexibility to work on this during FY26 Q2.

**Comment by JohnHalleyGotway** [comment]
Note that on Dec 8, 2025, @anewman89, @arbennett, and @JohnHalleyGotway met to discuss this issue (see [notes](https://docs.google.com/document/d/1r8ZysLrrgtArQmtuYsThEBMO0faEB2ep7-mOz59zOcA/edit?tab=t.0#heading=h.k9vscr2qsagi)). We were unsure how/where to add support for this in MET and METplus.

The information theory metrics are rather research-y and have certainly not gained wide adoption at operational centers. While it may not be used for day-to-day vx, it may prove useful to model developers.
- Should info theory metrics be a new STAT line type with full support added to Grid-Stat, Point-Stat, Stat-Analysis, and Series-Analysis - with output that's processed and plotted by METplus Analysis tools?
- Should it just be a new STAT-Analysis job computed from a time-series of MPR lines? If so, should that also go into Series-Analysis?
- Should it be implemented in METplus (not MET) via Python, somewhat similar to the Python code for computing indices on which Tina's worked?

We further discussed this issue during the METplus Land project meeting on December 11, 2025 with @anewman89, @KathrynNewman, @DanielAdriaansen, @georgemccabe, and @JohnHalleyGotway. We decided that Grid-Diag really is a good home for the mutual information metrics. We also note the Grid-Diag will be enhanced to write power sprectra in the next version. So it's a good home for distributional analysis that does not require strict matching of fcst/obs pairs.

For this work recommend:
1. Enhancing Grid-Diag to process an arbitrary number of masking regions rather than just one.
2. Enhancing Grid-Diag to make the computation of 1-D histograms, 2-D histograms, information theory, and power spectra configurable.

Note that we may want enhancements to METplus Analysis to plot the results.

**Comment by JohnHalleyGotway** [comment]
@anewman89, good news. I finally made some good progress on this issue:
- [x] Added support for multiple masking regions in the same run.
- [x] Restructured processing logic to make 1D histograms, 2D histograms, and info theory outputs optional.
- [x] Added basic Shannon entropy output computed for each variable/level and masking region.

Next steps are:
- [x] Add mutual information output corresponding to each 2D joint histogram.
- [x] Review progress with you to confirm it behaves and expected and identify areas for refinement and/or enhancement.

**Comment by anewman89** [comment]
Great, thanks for the update John! Looking forward to looking at some output with you!

**Comment by JohnHalleyGotway** [comment]
As discussed at the METplus Land project meeting on Jan 15, 2026, we should compare the entropy and mutual information output from Grid-Diag to values computed by another method.

Recommend running https://github.com/arbennett/HYEENNA for cross-validation. Note that I did try the [madvn/infotheory](https://github.com/madvn/infotheory) repo. While the Python package does install, it segfaults at runtime.

**Comment by JohnHalleyGotway** [comment]
I ran the following commands to validate the new Grid-Diag output.

1. Ran this updated [unit test](https://github.com/dtcenter/MET/blob/c359b9933c69d70982884aa3caca462f4b39ed6d/internal/test_unit/xml/unit_grid_diag.xml#L21) and used `ncdump` to note the following in the output file:
```
entropy_RH_Z2 = 5.722802, 6.356265, 5.705986 ;
entropy_PWAT_L0 = 5.571056, 4.596088, 5.524159 ;
mutual_information_RH_Z2_PWAT_L0 = 0.2446641, 0.4239301, 0.3418605 ;
```
2. Used the Python `scipy` and `sckit-learn` packages to compute these stats from the 1D and 2D histogram counts:
```
cd /d1/personal/johnhg/MET/MET_development/met-13.0.0/beta1/MET-feature_3171_mutual_information
python3

import netCDF4 as nc
from scipy.stats import entropy
from sklearn.metrics import mutual_info_score as mi

ds = nc.Dataset("test_output/grid_diag/grid_diag_out.nc")
hist1d = ds.variables["hist_RH_Z2"]
hist2d = ds.variables["hist_RH_Z2_PWAT_L0"]

# Entropy, using base 2 for "bits" units rather than the default natural log for "nats" units
entropy(hist1d[0], base=2), entropy(hist1d[1], base=2), entropy(hist1d[2], base=2)
# (5.722801898807304, 6.356264797434598, 5.705985771504789)

# Mutual information, converting the result from "nats" to "bits"
mi(None, None, contingency=hist2d[0]) * np.log2(np.e), mi(None, None, contingency=hist2d[1]) * np.log2(np.e), mi(None, None, contingency=hist2d[2]) * np.log2(np.e)
# (0.24466412613568606, 0.4239301292602261, 0.34186047086636684)
```
Confirm that those scores match exactly! Also note that I should specify the units as "bits" (rather than "nats") in the `units` attribute of the Grid-Diag `entropy_*` and `mutual_information_*` variables.

**Comment by anewman89** [comment]
This is awesome progress @JohnHalleyGotway!  If we can, adding [transfer entropy](https://en.wikipedia.org/wiki/Transfer_entropy) is a really nice next metric.  There are then conditional mutual information and conditional transfer entropy which we could explore in another issue if that is more appropriate.

**Comment by anewman89** [comment]
Actually, transfer entropy is going to be a problem for Grid-Diag as transfer entropy uses t-1 for conditional probabilities of Var X to Var Y.

**Comment by JohnHalleyGotway** [comment]
@anewman89, the Grid-Diag tool certainly does step through the input files one at a time. And users certainly can provide input files that comprise a time series. While Grid-Diag does track and report the min/max of the input timing information, it currently just aggregates all the input data into a single set of histograms.

We could modify the logic to create histograms separately for each input file to support computing transfer entropy and perhaps other stats. However, that's a non-trivial change and may require logic to check whether the inputs actually are a time series. We'd need to decide whether or not its worthwhile.

Grid-Diag is currently writing (1) entropy for 1d histograms and (2) mutual information for 2d histograms.
Adding [Joint Entropy](https://en.wikipedia.org/wiki/Joint_entropy) for the 2d histogram would be very easy. So let me know if you like me to add that for this PR.

Otherwise, I'd recommend proceeding with the review of PR #3316. And we can discuss whether the bigger change to the processing logic to compute transfer entropy is warranted in another issue/PR.

**Comment by anewman89** [comment]
Thanks John.  Lets add joint entropy as you suggest since it doesn't modify more of Grid-Diag.  Then we can call this PR done and move forward with a discussion on transfer entropy.

**Comment by arbennett** [comment]
Great to see progress on this! One quick clarification is that transfer entropy is actually just a variant of conditional mutual information, so if you do CMI you get TE for free - it's just a matter of conditioning on a time lag vs another variable, which may inform which package it ends up in.

**Comment by anewman89** [comment]
Thanks Andrew! I should have thought about that more, that makes total sense!  From a code standpoint, it's probably easier to add a third variable from the same time step for CMI rather than previous time steps for TE.  That being said, both are probably better in another issue and PR.

**Comment by JohnHalleyGotway** [comment]
Added joint entropy to the [feature_3171_mutual_information](https://github.com/dtcenter/MET/tree/feature_3171_mutual_information) branch. Here's the newly generated output from MET:
```
joint_entropy_RH_Z2_PWAT_L0 = 11.04919, 10.52842, 10.88828 ;
```

And here's Python code to validate this output:
```
cd /d1/personal/johnhg/MET/MET_development/met-13.0.0/beta1/MET-feature_3171_mutual_information
python3

import netCDF4 as nc
from scipy.stats import entropy

ds = nc.Dataset("test_output/grid_diag/grid_diag_out.nc")
hist2d = ds.variables["hist_RH_Z2_PWAT_L0"]

# Entropy, using base 2 for "bits" units rather than the default natural log for "nats" units
entropy(hist2d[0].ravel()/hist2d[0].sum(), base=2), entropy(hist2d[1].ravel()/hist2d[1].sum(), base=2), entropy(hist2d[2].ravel()/hist2d[2].sum(), base=2)
# (11.049194074343399, 10.528422981240544, 10.88828406597514)
```

This Python and MET joint entropy values match.
